src/hdmf/backends/hdf5/h5tools.py

```diff
@@ -1099,6 +1100,8 @@ def write_dataset(self, **kwargs):  # noqa: C901
             dataio = data
             link_data = data.link_data
             data = data.data
+        if isinstance(data, TermSetWrapper):
+            data = data.item
```

Comment:
I think you will need to do the same in ``set_attributes`` https://github.com/hdmf-dev/hdmf/blob/a0fdb244727b622de17dc470eec9bc5d1252a510/src/hdmf/backends/hdf5/h5tools.py#L929-L931

at the beginning of the loop do ``isinstance(data, TermSetWrapper):  value = value.item``